pkg/backend/httpstate/state.go

```diff
@@ -179,20 +179,20 @@ func (u *cloudUpdate) RecordAndDisplayEvents(
 	displayEvents := make(chan engine.Event)
 	go display.ShowEvents(label, action, stackRef.Name(), op.Proj.Name, displayEvents, innerDone, opts, isPreview)
 
-	seen := make(map[resource.URN]engine.StepEventMetadata)
-
 	// We maintain a sequence counter for each event to ensure that the Pulumi Service can
 	// ensure events can be reconstructured in the same order they were emitted. (And not
 	// out of order from parallel writes and/or network delays.)
 	eventIdx := 0
-	for e := range events {
-		eventIdx++
 
-		// First echo the event to the local display.
-		displayEvents <- e
+	// We start the requests to record engine events in separate Go routines since they can
+	// all be done independently, and updates with a "chatty" event stream can have serious
+	// perf problems when issuing the requests serially.
+	var wg sync.WaitGroup
+	recordEngineEvent := func(event engine.Event, eventIdx int) {
+		wg.Add(1)
+		defer func() { wg.Done() }()
```

Comment:
Can you do just `defer wg.Done()` instead?

Reply:
Absolutely! That was muscle memory... Done.